docs/Client.md

```diff
@@ -0,0 +1,928 @@
+
+```js
+'use strict'
+
+const { createServer } = require('http')
+const undici = require('undici')
+const fastify = require('fastify')
+
+const nodeServer = createServer((request, response) => {
+  response.end('Hello, World! From Node.js HTTP Server')
+})
+
+const fastifyServer = fastify()
+
+nodeServer.listen(0, () => {
+  console.log('Node Server listening')
+
+  const client = new undici.Client(`http://localhost:${nodeServer.address().port}`)
+
+  fastifyServer.route({
+    url: '/',
+    method: 'GET',
+    handler: (request, response) => {
+      client.stream({
+        path: '/',
+        method: 'GET',
+        opaque: response
+      }, ({ opaque }) => opaque.raw)
+    }
+  })
+
+  fastifyServer.listen(0, () => {
+    console.log('Fastify Server listening')
+    fastifyServer.inject({
+      path: '/',
+      method: 'GET'
+    }).then(({ statusCode, body }) => {
+      console.log(`response received ${statusCode}`)
+      console.log('body:', body)
+
+      client.close()
+      fastifyServer.close()
+      nodeServer.close()
+    })
+  })
+})
+```
+
+### `Client.upgrade(options[, callback])`
+
+Upgrade the client to a different protocol.
+
+Arguments:
+
+* **options** `UpgradeOptions`
+  
+* **callback** `(error: Error | null, data: UpgradeData) => void` (optional) - If a callback is not provided, a promise is returned
+
+Returns: `void | Promise<UpgradeData>` - Only returns a `Promise` if no `callback` argument was passed
+
+#### Parameter: `UpgradeOptions`
+
+* **path** `string`
+* **method** `string` (optional) - Default: `'GET'`
+* **headers** `UndiciHeaders` (optional) - Default: `null`
+* **protocol** `string` (optional) - Default: `'Websocket'` - A string of comma separated protocols, in descending preference order.
+* **signal** `AbortSignal | EventEmitter | null` (optional) - Default: `null`
+
+#### Parameter: `UpgradeData`
+
+* **headers** `http.IncomingHeaders`
+* **socket** `stream.Duplex`
+* **opaque** `unknown`
+
+#### Example 1 - Basic Upgrade Request
+
+```js
+'use strict'
+const { Client } = require('undici')
+const { createServer } = require('http')
+
+const server = createServer((request, response) => {
+  response.statusCode = 101
+  response.setHeader('connection', 'upgrade')
+  response.setHeader('upgrade', request.headers.upgrade)
+  response.end()
+})
+
+server.listen(() => {
+  const client = new Client(`http://localhost:${server.address().port}`)
+
+  client
+    .upgrade({ path: '/' })
+    .then(({ headers, socket }) => {
+      socket.on('end', () => {
+        console.log(`upgrade: ${headers.upgrade}`)
+        client.close()
+        server.close()
+      })
+      socket.end()
+    })
```

Comment:
what happens if this rejects? Maybe add a catch?

Reply:
Good point, will do! (Best to not have unhandeledRejections laying around 😉 )